Close 4588 - Enable auto-compatibility XL WooCommerce Sales Triggers with DelayJS

iCaspar opened the pull request:
## Description

This adds a new ThirdParty compatibility to correctly exclude the XL WooCommerce Sales Triggers combined js and the dependent jQuery on WooCommerce single product pages.

Fixes #4588

## Type of change

Please delete options that are not relevant.

- [X] Enhancement (non-breaking change which improves an existing functionality)

## Is the solution different from the one proposed during the grooming?

Yes. Initial grooming plan was to simply add the sales triggers script to the default exclusions list. However, the script is dependent on jQuery, which requires delaying that also, but ONLY on single WC product pages.

## How Has This Been Tested?
- [X] Integration test and fixtures are added for new expected behavior
- [X] Confirmed that scripts are excluded as expected on Product pages, but not on other (homepage) in local environment.

# Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

iCaspar committed:
Add integration test for new compat subscriber

Add the integration test and fixture for a new subscriber to handle XLWoo Sales Triggers plugin compatibility.

Given the plugin is active, we have 2 cases:
1. When we are on a single product page, we add the plugin's script + jquery patterns to delayJS exclusions.
2. When we are not on a single product page, we don't change any delayJS exclusions.

iCaspar committed:
Add compatibility subscriber and wire it up

The compatibiity subscriber is added that will pass the integration tests added previously.
The new subscriber is added to the ThirdParty ServiceProvider.
The new subscriber is added to Rocket's common_subscribers array.

iCaspar committed:
Fix phpcs

Adjust some spacing.
Add array commas.
Add @param to docblock.

iCaspar committed:
Fix classname case mismatch in service provider

remyperona commented on inc/ThirdParty/Plugins/Ecommerce/XLWooCommerceSalesTriggersSubscriber.php:
```diff
@@ -0,0 +1,67 @@
+				[
+					'xl-woocommerce-sales-triggers/assets/js/wcst_combined.min.js',
+					'/jquery-?[0-9.](.*)(.min|.slim|.slim.min)?.js',
+					'js-(before|after)',
+					'(?:/wp-content/|/wp-includes/)(.*)',
```
are those also needed for the compatibility?

iCaspar replied:
Nope!

remyperona commented on inc/ThirdParty/Plugins/Ecommerce/XLWooCommerceSalesTriggersSubscriber.php:
```diff
@@ -0,0 +1,67 @@
+			return $exclusions;
+		}
+
+		if ( 'product' !== get_post_type() || ! is_single() ) {
```
In the WC context, you can use `is_product()` directly instead

iCaspar replied:
Good point.

iCaspar committed:
Changes per code review suggestions

1. Only delay jQuery, not the whole wp-content things.
2. Check for product type with is_product() directly.

remyperona commented on inc/ThirdParty/Plugins/Ecommerce/XLWooCommerceSalesTriggersSubscriber.php:
```diff
@@ -0,0 +1,65 @@
+			return $exclusions;
+		}
+
+		if ( ! is_product() || ! is_single() ) {
```
won't the `is_single()` exclude jquery on any post type single page?

iCaspar replied:
The logic here is, we're going to bail out whenever it's NOT a product OR whenever it's NOT single. So, I think, checking against not/OR in both sides will allow it to add the exclusions only when it IS a product AND IS single.

iCaspar replied:
In other words, when it is not a product, it will meet the condition before it ever checks for single -- so it will have bailed out and not affect singles of other types.

remyperona replied:
Checking back, `is_single()` only applies to the `post` post type, not all of them (`is_singular()` does).

Also, the underlying code for `is_product()` is `is_singular( array( 'product' ) )` so in the end the "single" check is already done at that point.

So I would suggest only keeping `is_product()` there, as it's going to check everything needed, what do you think?

iCaspar replied:
Indeed, that will work. Thanks.

iCaspar committed:
Use single post instead of homepage testcase.

Question arose during review about whether bail-early condition would add jQuery for other single entries besides products.

This modifies the testcase for not adding exclusions case to verify specifically with a single-not-product item to confirm that behavior is as expected.

iCaspar committed:
Remove redundant check for singular.

is_product() already checks for is_singular(), so we're removing the check for is_single()

Mai-Saad commented:
Couldn't reproduce the main problem on the test site, however, please find the notes below:
- Using the test template (not product page), can see a jQuery console error while Defer JS is enabled => on both trunk and branch (this error is not in nowprocket page)
- Q: do we need to exclude anything for` jquery-migrate`?
@Tabrisrp @piotrbak  what do you think? do we need to do further change(s) here?

Note for retest:
The script `/wp-content/plugins/xl-woocommerce-sales-triggers/assets/js/wcst_combined.min.js` is only delayed on the product page now (thanks to @alfonso100 )

piotrbak commented:
@Mai-Saad I think we need to reproduce the issue to move forward. For future exclusions, we need to rely on product pages and environemnts, not the templates.
